telepresence/proxy/operation.py

```diff
@@ -0,0 +1,312 @@
+                self.intent.name,
+                dict(telepresence=runner.session_id),
+                dict(telepresence=runner.session_id),
+                {p: p
+                 for p in self.intent.expose.remote()},
+            )
+            self.manifests.append(svc)
+
+        self.remote_info = make_remote_info_from_pod(pod)
+
+    def act(self, runner: Runner) -> RemoteInfo:
+        runner.show(
+            "Starting network proxy to cluster using "
+            "new Pod {}".format(self.intent.name)
+        )
+        create_with_cleanup(runner, self.manifests)
+
+        wait_for_pod(runner, self.remote_info)
+
+        return self.remote_info
+
+
+def ensure_correct_version(runner: Runner, remote_info: RemoteInfo) -> None:
+    """
+    Ensure remote container is running same version as we are
+    """
+    remote_version = remote_info.remote_telepresence_version()
+    if remote_version != image_version:
+        runner.write("Pod is running Tel {}".format(remote_version))
+        raise runner.fail((
+            "The remote datawire/telepresence-k8s container is " +
+            "running version {}, but this tool is version {}. " +
+            "Please make sure both are running the same version."
+        ).format(remote_version, image_version))
+
+
+def set_expose_ports(
+    expose: PortMapping, pod: Manifest, container_name: str
+) -> None:
+    """
+    Merge container ports into the expose list
+    """
+    containers = pod["spec"]["containers"]  # type: Iterable[Manifest]
+    for container in containers:
+        if not container_name or container["name"] == container_name:
+            expose.merge_automatic_ports([
+                port["containerPort"] for port in container.get("ports", [])
+                if port["protocol"] == "TCP"
+            ])
+            break
+
+
+class Swap(ProxyOperation):
+    def prepare(self, runner: Runner) -> None:
+        self.manifests = []  # type: List[Manifest]
+
+        # Grab original deployment's Pod Config
+        deployment = get_deployment(runner, self.intent.name)  # type: Manifest
+        self.deployment_type = deployment["kind"]  # type: str
+        self.original_replicas = deployment["spec"]["replicas"]  # type: str
+
+        # Compute a new name that isn't too long, i.e. up to 63 characters.
+        # https://github.com/kubernetes/community/blob/master/contributors/design-proposals/architecture/identifiers.md
+        new_pod_name = "{name:.{max_width}s}-{id}".format(
+            name=self.intent.name,
+            id=runner.session_id,
+            max_width=(50 - (len(runner.session_id) + 1))
+        )
+
+        # Perform the relevant swap changes to the pod spec
+        pod_spec = deployment["spec"]["template"]["spec"]
+        pod_spec["restartPolicy"] = "Never"
+        if self.intent.service_account:
+            pod_spec["serviceAccount"] = self.intent.service_account
+
+        # Find the relevant container
+        if self.intent.container:
```

Comment:
Nope! Good catch. The legacy code just crashes, but at least with a useful error message. This would have crashed with a NameError. Anyhow, now we fail with a similar useful error message rather than crashing, at least in the new code path.